repo2docker/app.py

```diff
@@ -605,12 +605,24 @@ def wait_for_container(self, container):
         try:
             for line in container.logs(stream=True):
                 self.log.info(line.decode("utf-8"), extra=dict(phase="running"))
+
         finally:
             container.reload()
             if container.status == "running":
                 self.log.info("Stopping container...\n", extra=dict(phase="running"))
                 container.kill()
             exit_code = container.attrs["State"]["ExitCode"]
+
+            container.wait()
+
+            self.log.info(
+                "Container finished running.\n".upper(), extra=dict(phase="running")
+            )
+            # are there more logs? Let's send them back too
+            late_logs = container.logs().decode("utf-8")
+            for line in late_logs.split("\n"):
```

Comment:
In my runs, late_logs is never empty, it's always the complete logs of the container. So this is always duplicating the logs and *also* missed the same logs.

Reply:
Absolutely right. I added this thinking that maybe we missed some logs. But that isn't the problem.

If I remove/comment out the "delete stopped container" part of r2d and then run `docker logs <container>` afterwards the logs are also empty (when this problems occurs). So I think your ideas about `PYTHONUNBUFFERED` and `stdbuf` are the way to go investigating.

I think, the "late log" lines should be removed